refactor: Kafka Webhook Management

gavindsouza opened the pull request.

gavindsouza committed:
perf: Don't re-fetch frappe.utils context

Set doc as None, and pass the doc on demand instead of duplicate calls
in each thread. Maintain as global variable instead, much faster -
specifically 100 microseconds to 85 nanoseconds ^_^

gavindsouza committed:
refactor: KafkaHook.run_webhooks

* Break down and separate functions based on actions + lesser indents!
* Rename functions to denote explicit behaviour
* Use typing for better understandability

gavindsouza committed:
perf(enqueue_webhook): Use cached Kafka Hook doc

Given how frequently this function is called, better to rely on Redis
than Database for IO + marshalling. Better yet we should navigate a way
to skip building the Document object altogether since that's unnecessary

gavindsouza committed:
refactor: Share a single Kafka Producer between threads

We're maintaining one client that may be shared between threads of the
same process, based on the documentation of the KafkaProducer client:

    The producer is thread safe and sharing a single producer instance across
    threads will generally be faster than having multiple instances.

    The producer consists of a pool of buffer space that holds records that
    haven't yet been transmitted to the server as well as a background I/O
    thread that is responsible for turning these records into requests and
    transmitting them to the cluster.

The clients generated will be stored under the global bulkwebhook.PRODUCER_MAP
separated by site and kafka settings

sagarvora commented on bulkwebhook/bulk_webhook/doctype/kafka_hook/kafka_hook.py:
```diff
@@ -7,9 +7,11 @@
 from bulkwebhook.bulk_webhook.doctype.bulk_webhook.bulk_webhook import log_request
+from frappe.integrations.doctype.webhook.webhook import get_context as get_webhook_context
+
+WEBHOOK_CONTEXT = get_webhook_context(doc=None)
```
let's just hardcode it IMO. maybe the context in webhook changes and contains values that cannot be set as a global?

also the `get_safe_globals` call access objects that need a corresponding local (e.g. `frappe.db`) - this may fail if the local isn't initialized 😅 

we just need the utils anyway. perhaps it's best to get them directly from `safe_exec.add_data_utils`?

gavindsouza replied:
Yeah, adding utils directly is better 😄 - will do that!

sagarvora commented on bulkwebhook/bulk_webhook/doctype/kafka_hook/kafka_hook.py:
```diff
@@ -42,21 +44,18 @@ def validate_condition(self):
-
-def enqueue_webhook(doc, kafka_hook):
-    hook = frappe.get_doc("Kafka Hook", kafka_hook.get("name"))
+def enqueue_webhook(doc: Document, kafka_hook: frappe._dict):
```
Looks like we're not calling enqueue here 😅
`run_kafka_hook`?

sagarvora commented on bulkwebhook/bulk_webhook/doctype/kafka_settings/kafka_settings.py:
```diff
@@ -1,9 +1,12 @@
 
 class KafkaSettings(Document):
-    pass
+    def on_update(self):
```
`on_trash` as well? or we can just extend the `clear_cache` method.

gavindsouza replied:
~Would've liked to use https://github.com/frappe/frappe/pull/17012 instead. But think I'll just add on_trash too for simplicity~

Extending `clear_cache` seems like a much better option - thanks!

sagarvora commented on bulkwebhook/bulk_webhook/doctype/kafka_hook/kafka_hook.py:
```diff
@@ -7,9 +7,11 @@
-from frappe.utils.safe_exec import get_safe_globals
 from bulkwebhook.bulk_webhook.doctype.kafka_settings.kafka_utlis import send_kafka
 from bulkwebhook.bulk_webhook.doctype.bulk_webhook.bulk_webhook import log_request
+from frappe.integrations.doctype.webhook.webhook import get_context as get_webhook_context
```
If I'm not mistaken imports are in a frappe-first sequence 😅

gavindsouza replied:
Is this a convention for the project? The order doesn't really matter unless imports are circular, something is monkey patched, or another requirement, no? 

Anyway, eliminating this line altogether 😄

sagarvora replied:
Yeah, seems like a convention.

We do it for India Compliance as well. For custom apps, frappe imports can be thought of as "standard" imports, so I guess it's nice to have a differentiation.

ankush replied:
Just `isort` and forget :smile:

gavindsouza committed:
perf: Generate data utils dict directly instead of get_context

Further improves performance changes indtroduced in 2846b4939e1d68b8f98fc56909af014a1c0a33ca and bypasses possible unintended failures described in https://github.com/aakvatech/Bulk-Webhook/pull/1#discussion_r1013705470

gavindsouza committed:
refactor: enqueue_webhook => run_kafka_hook

Corrected improperly named function to match actions

gavindsouza committed:
style(doctype[dir]): Black

gavindsouza committed:
fix(KafkaSettings): Extend clear_cache to evict stale producers

sagarvora commented on bulkwebhook/bulk_webhook/doctype/kafka_hook/kafka_hook.py:
```diff
@@ -7,11 +7,21 @@
+    return data_utils
+
+
+WEBHOOK_CONTEXT = get_safe_frappe_utils()
```
```suggestion
WEBHOOK_CONTEXT = {"utils": get_safe_frappe_utils()}
```

gavindsouza replied:
reliable reviewers & a solid CI run big projects 🙏🏼

sagarvora commented:
Rest looks good ✅ 

Maybe we can just test it once.

gavindsouza committed:
fix(WEBHOOK_CONTEXT): Namespaced frappe utils

Co-authored-by: Sagar Vora <sagar@resilient.tech>

sagarvora commented on bulkwebhook/bulk_webhook/doctype/kafka_hook/kafka_hook.py:
```diff
@@ -7,11 +7,21 @@
+def get_safe_frappe_utils():
+    from frappe.utils.safe_exec import add_data_utils
+
+    data_utils = {}
```
Must be dottable: `_dict`

sagarvora commented on bulkwebhook/bulk_webhook/doctype/kafka_hook/kafka_hook.py:
```diff
@@ -98,62 +140,31 @@ def run_webhooks(doc, method):
 
         if trigger_webhook and event and webhook.webhook_docevent == event:
-            _webhook_request(webhook)
+            if webhook.name not in frappe.flags.kafkahook_executed.get(doc.name, []):
```
- Can we put this first? Feels like we should run eval later.
- Flat > Nested.

gavindsouza replied:
Agreed. Had overlooked this bit because it didn't seem like a "big enough" problem in comparison. Updated via https://github.com/aakvatech/Bulk-Webhook/pull/1/commits/9b932d271c9b584163de56f3141ddaf5bbabeb71

gavindsouza committed:
perf: eager? evaluation of condiitons in KafkaHook.run_webhooks

Eliminate blocks of code as early as possible by least amount of
computation

aakvatech committed:
Merge branch 'main' into refactor-kafka_conn
